plugins/index-by-search/src/main/java/org/elasticsearch/plugin/indexbysearch/AbstractAsyncScrollAction.java

```diff
@@ -0,0 +1,339 @@
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.plugin.indexbysearch;
+
+import static java.lang.Math.max;
+import static java.util.Collections.unmodifiableList;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicReference;
+
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.ActionRequest;
+import org.elasticsearch.action.bulk.BulkItemResponse;
+import org.elasticsearch.action.bulk.BulkItemResponse.Failure;
+import org.elasticsearch.action.bulk.BulkRequest;
+import org.elasticsearch.action.bulk.BulkResponse;
+import org.elasticsearch.action.bulk.TransportBulkAction;
+import org.elasticsearch.action.index.IndexResponse;
+import org.elasticsearch.action.search.ClearScrollRequest;
+import org.elasticsearch.action.search.ClearScrollResponse;
+import org.elasticsearch.action.search.SearchRequest;
+import org.elasticsearch.action.search.SearchResponse;
+import org.elasticsearch.action.search.SearchScrollRequest;
+import org.elasticsearch.action.search.TransportClearScrollAction;
+import org.elasticsearch.action.search.TransportSearchAction;
+import org.elasticsearch.action.search.TransportSearchScrollAction;
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.logging.ESLogger;
+import org.elasticsearch.common.unit.ByteSizeValue;
+import org.elasticsearch.plugin.indexbysearch.AbstractAsyncScrollAction.AsyncScrollActionRequest;
+import org.elasticsearch.search.SearchHit;
+
+/**
+ * Abstract base for scrolling across a search and executing bulk actions on all
+ * results. Right now that is only used by AsyncIndexBySearchAction but its
+ * still nice because this handles the scrolling while AsyncIndexBySearchAction
+ * handles the document building.
+ */
+public abstract class AbstractAsyncScrollAction<Request extends ActionRequest<?> & AsyncScrollActionRequest, Response> {
+    public interface AsyncScrollActionRequest {
+        /**
+         * Maximum number of processed documents. -1 means process all
+         * documents.
+         */
+        int size();
+        /**
+         * Should vesion conflicts be saved to the failures list?
+         */
+        boolean saveVersionConflicts();
+        /**
+         * How many failures must be accumulated before the operation aborts?
+         * Defaults to 1. This action may accumulate more than this many
+         * failures because failure is checked after each bulk batch.
+         */
+        int failuresCauseAbort();
```

Comment:
I suppose so. We couldn't store all the failures if we used a ratio though. Its certainly a solvable problem.